### ISSUE TITLE
Add optical flow baseline

### ISSUE THREAD

**Issue by jacobbieker** [openclimatefix/satflow#47]

**Comment by jacobbieker** [comment]
The baseline is running, but using the HRV channel to compute optical flow (the cloud masks didn't work for computing flow), and then applying the flow to the cloud masks to predict the next 48 timesteps (4 hours) results in a MSE of 0.1541 vs a naive baseline of just using the current frame for all future frame MSE loss of 0.1566. So optical flow gives some improvement, but not a ton. Most of the current networks being trained have MSE of around 0.1-0.07~8 so theoretically are already doing better than optical flow or the naive baseline

**Comment by JackKelly** [comment]
That's really exciting that the neural networks are doing better than optical flow!

My guess would be that optical flow does really well over short time horizons (30 minutes?  1 hour?).  It might be interesting to separately measure error across different forecast horizons (e.g. 30 mins ahead, 1 hr ahead, 2 hrs ahead, and 4 hrs ahead).  Some users might be particularly interested in very short horizon forecasts.

**Comment by jacobbieker** [comment]
Yeah! It saves out the mean error for each timestep, so I do have that information for every 5 minutes once it finishes, its just somewhat slow and running in the background. I'm going to try adding in the error per timestep to the NN outputs soon too, so its easier to compare.

**Comment by jacobbieker** [comment]
The baseline finished, but gives some kinda odd results. As you can see in the plot, the MSE loss actually decreases as time go on, except for the last timestep for some reason. If you take the last timestep as the "prediction" for all of the frames, the same thing occurs in reverse, except the final loss is 0 as the images are identical. Probably missing something or some bug, but it doesn't quite make sense to me.

![MSE_Loss](https://user-images.githubusercontent.com/7170359/125455029-293fcbc4-b642-40e2-aba3-6d3b6d3f448c.png)

**Comment by jacobbieker** [comment]
It seems up until about ~1hr in the future, the optical flow and baseline give better results than current models do.
![MSE_Loss](https://user-images.githubusercontent.com/7170359/125460964-7be7fc41-5067-4327-afa0-e7c54a446358.png)

**Comment by jacobbieker** [comment]
Overall, using the current frame as the output results in a MSE of 0.160, using optical flow gives MSE of 0.157. One of the ConvLSTM models that is training gives a validation MSE of 0.0712 overall, with a MSE of 0.036 for the first future frame, and 0.095 for the last frame, suggesting that the ML models already heavily outperform the baseline models already. Still need work on finding possibly better metrics than MSE which might change the comparison, but at least for now it is looking pretty good I think

**Comment by JackKelly** [comment]
Sounds extremely promising!

**Comment by jacobbieker** [comment]
Next up will be redoing it with the actual satellite channels, as the binary masks might not be the best way forward #6 The ML models trained to produce the 12 satellite channels tend to have similar loss as for the binary masks, so if that holds for optical flow as well, then the results might be quite similar, which is a good sign for using ML for this!

**Comment by jacobbieker** [comment]
Optical flow baseline is running now for the 12 sat channels, where the flow is determined for each channel separately. Preliminary results is that the MSE is ~1.05 for optical flow. For models I've been training, their loss tends to be ~0.6-0.3, so the models are potentially much better than optical flow for the actual channels compared to the binary cloud masks!

**Comment by jacobbieker** [comment]
Plots from the running job. The MSE depends a lot on the specific channel for optical flow. There also seems to be large jumps for certain channels after an hour, which I am not entirely sure why that happens.

![mse_per_channel](https://user-images.githubusercontent.com/7170359/127460256-9408d87f-d2bf-4035-82b7-2a67c58c9387.png)
![mse_per_timestep](https://user-images.githubusercontent.com/7170359/127460263-97966436-4a25-4094-92b0-c521cb9f0ad5.png)
![mse_IR120](https://user-images.githubusercontent.com/7170359/127460250-8364d1f2-3b7e-4cf0-839b-d4e331f18a12.png)
![mse_IR134](https://user-images.githubusercontent.com/7170359/127460252-7f62d7fc-6c6b-4304-9f49-5d73db4ea828.png)
![mse_VIS006](https://user-images.githubusercontent.com/7170359/127460267-3b596ccc-c0cc-4a90-87dc-bb6e8e665366.png)
![mse_VIS008](https://user-images.githubusercontent.com/7170359/127460272-75091a23-1a2b-4c4a-bd54-9aa87687fde0.png)
![mse_WV062](https://user-images.githubusercontent.com/7170359/127460276-5e8caa1f-5b72-4a8b-a24d-c8d5456821db.png)
![mse_WV073](https://user-images.githubusercontent.com/7170359/127460279-b2014ab5-fd00-4b5c-8935-2b4ea1b3d6c9.png)
![mse_HRV](https://user-images.githubusercontent.com/7170359/127460197-3290baca-0076-4420-9ceb-2568a93124a5.png)
![mse_IR016](https://user-images.githubusercontent.com/7170359/127460228-728fa47f-53cc-47a1-acb2-16cbaff0c9df.png)
![mse_IR039](https://user-images.githubusercontent.com/7170359/127460232-090aff4e-ee7d-4122-88a5-1696a0a7788b.png)
![mse_IR087](https://user-images.githubusercontent.com/7170359/127460236-1ec82243-ac62-47b6-b90e-1924cda1a921.png)
![mse_IR097](https://user-images.githubusercontent.com/7170359/127460241-35ebcd70-cce6-436e-9716-fcecaf2b1688.png)
![mse_IR108](https://user-images.githubusercontent.com/7170359/127460245-db0ea196-7f5d-469f-ada6-3569fbb2356b.png)

**Comment by jacobbieker** [comment]
The timedelta between all the future timesteps is 5min, like it should be, so not sure for the jump after timestep ~7, but it doesn't seem to be an issue with the dataloader

**Comment by jacobbieker** [comment]
Things to try: Take the mean of a few pairs of previous images to do the optical flow, and get the optical flow between the last frame and the previous, previous frame, to see if extra movement helps on especially channels that are a bit blurrier and probably harder to compute the flow of

**Comment by JackKelly** [comment]
Awesome work!

Looking again at the plots, I am really surprised that the MSE for the 'current image' and 'optical flow' are basically identical.  If I've understood correct then, if optical flow is doing anything sane, then it really should out-perform 'current image', at least for the first few timesteps of the prediction.  But maybe the issue is that I haven't got a good intuitive feel for what MSE of a pair of images means :)

**Comment by jacobbieker** [comment]
Yeah, I am too. Computing the Optical flow as the average of the optical flow of the last 5 pairs of images gives more of a difference for the MSE.
![mse_HRV](https://user-images.githubusercontent.com/7170359/128509821-b44e7e70-88d1-490d-8243-81ead612ac8c.png)
![mse_IR016](https://user-images.githubusercontent.com/7170359/128509823-ba6be5bf-4e15-4911-9351-9dd52f83d55c.png)
![mse_IR039](https://user-images.githubusercontent.com/7170359/128509825-d0b986b9-5bc4-47e8-be88-f6658f9c4dc5.png)
![mse_IR087](https://user-images.githubusercontent.com/7170359/128509827-8a0e91db-d0e2-41aa-a714-4b6f90028341.png)
![mse_IR097](https://user-images.githubusercontent.com/7170359/128509829-561d58d8-0ef5-481a-aed1-3e4011c79766.png)
![mse_IR108](https://user-images.githubusercontent.com/7170359/128509830-73ac1513-5373-4506-b5a1-4535ec00cd46.png)
![mse_IR120](https://user-images.githubusercontent.com/7170359/128509831-84041e24-ae1c-4bd0-ae87-7b018293adcf.png)
![mse_IR134](https://user-images.githubusercontent.com/7170359/128509832-5ae32bc7-3370-4507-ad79-92b3bf23095e.png)
![mse_per_channel](https://user-images.githubusercontent.com/7170359/128509833-c62bacff-aaf3-4282-8f51-4a95145c51b3.png)
![mse_per_timestep](https://user-images.githubusercontent.com/7170359/128509836-133f54f9-fc76-4a4a-a1a2-2e3d9ab30d3f.png)
![mse_VIS006](https://user-images.githubusercontent.com/7170359/128509837-e21849cf-2e49-4cdc-b98d-6d6cf6c08a44.png)
![mse_VIS008](https://user-images.githubusercontent.com/7170359/128509839-63708e38-e842-48d9-b66d-8a6b4f45c4cf.png)
![mse_WV062](https://user-images.githubusercontent.com/7170359/128509840-aaf8c345-fb79-4695-99bd-2397df04c005.png)
![mse_WV073](https://user-images.githubusercontent.com/7170359/128509845-e198111a-7573-4d0c-9fb5-098d0942db7e.png)
